pulp_docker/app/views.py

```diff
@@ -14,3 +57,235 @@ def get(self, request):
+
+        Args:
+            path (str): The fully qualified path to the file to be served.
+
+        Returns:
+            HttpResponse: A response with X-Accel-Redirect header.
+        """
+        response = HttpResponse()
+        response['X-Accel-Redirect'] = path
+        return response
+
+    def _redirect(self, request):
+        """
+        Get redirect-to-streamer response.
+
+        Args:
+            request (django.http.HttpRequest): A request for a content artifact.
+
+        Returns:
+            HttpResponseRedirect: Redirect to streamer.
+        """
+        redirect = settings.CONTENT['REDIRECT']
+
+        enabled = redirect['ENABLED']
+        host = redirect['HOST'] or request.get_host().split(':')[0]
+        port = redirect['PORT'] or request.get_port()
+        prefix = redirect['PATH_PREFIX']
+
+        if not enabled:
+            return HttpResponseNotFound()
+
+        path = os.path.join(
+            prefix.strip('/'),
+            self._published_path(request)
+        )
+
+        url = '{scheme}://{host}:{port}/{path}'.format(
+            scheme=request.scheme,
+            host=host,
+            path=path,
+            port=port)
+
+        log.debug(_('Redirected: %(u)s'), {'u': url})
+        response = HttpResponseRedirect(url)
+        return response
+
+    def _dispatch(self, path, headers):
+        """
+        Dispatch to the appropriate responder (method).
+
+        Args:
+            path (str): The fully qualified path to the file to be served.
+
+        Returns:
+            django.http.StreamingHttpResponse: on found.
+            django.http.HttpResponseNotFound: on not-found.
+            django.http.HttpResponseForbidden: on forbidden.
+            django.http.HttpResponseRedirect: on redirect to the streamer.
+        """
+        server = settings.CONTENT['WEB_SERVER']
+
+        try:
+            responder = self.RESPONDER[server]
+        except KeyError:
+            raise ValueError(_('Web server "{t}" not supported.').format(t=server))
+        else:
+            return responder(self, path, headers)
+
+    # Mapping of responder-method based on the type of web server.
+    RESPONDER = {
+        'django': _django,
+        'apache': _apache,
+        'nginx': _nginx,
+    }
+
+
+class TagView(views.APIView, ServeContentMixin):
+    """
+    APIView for Docker Registry v2 tag API
+    """
+
+    authentication_classes = []
+    permission_classes = []
+    content_negotiation_class = IgnoreClientContentNegotiation
+
+    def get(self, request, path, tag_name):
+        """
+        Return a response to the "GET" action.
+        """
+        distribution = get_object_or_404(DockerDistribution, base_path=path)
+        try:
+            ca = ContentArtifact.objects.get(
+                content__in=distribution.publication.repository_version.content,
+                relative_path=tag_name)
+            content = ca.content.cast()
+            if content.manifest:
```

Comment:
I can't remember-- do we allow a single Tag object to reference both a Manifest and a Manifest List? If so, this could have unexpected results.

Reply:
A Tag should only reference one or the other.